File: src-tauri/src/proxy.rs
use std::collections::HashMap;
use std::convert::Infallible;
use std::net::SocketAddr;
use std::sync::Arc;
use tokio::sync::Mutex;
use warp::{Filter, Reply};
use reqwest::Client;
use bytes::Bytes;

#[derive(Clone)]
pub struct ProxyServer {
    port: u16,
    client: Client,
}

impl ProxyServer {
    pub fn new(port: u16) -> Self {
        let client = Client::builder()
            .redirect(reqwest::redirect::Policy::limited(3))
            .timeout(std::time::Duration::from_secs(60))
            .connect_timeout(std::time::Duration::from_secs(10))
            // For development: accept invalid certificates
            // TODO: Remove this in production or make it configurable
            .danger_accept_invalid_certs(true)
            .build()
            .unwrap_or_else(|_| Client::new());

        ProxyServer { port, client }
    }

    pub async fn start(self) -> Result<(), Box<dyn std::error::Error>> {
        let addr: SocketAddr = ([127, 0, 0, 1], self.port).into();
        
        let proxy_server = Arc::new(Mutex::new(self));
        
        // OpenAI proxy route
        let openai_route = warp::path!("api" / "openai" / ..)
            .and(warp::method())
            .and(warp::path::full())
            .and(warp::query::<HashMap<String, String>>())
            .and(warp::header::headers_cloned())
            .and(warp::body::bytes())
            .and(with_proxy(proxy_server.clone()))
            .and_then(handle_proxy_request);

        // Azure proxy route
        let azure_route = warp::path!("api" / "azure" / ..)
            .and(warp::method())
            .and(warp::path::full())
            .and(warp::query::<HashMap<String, String>>())
            .and(warp::header::headers_cloned())
            .and(warp::body::bytes())
            .and(with_proxy(proxy_server.clone()))
            .and_then(handle_proxy_request);

        // Google proxy route
        let google_route = warp::path!("api" / "google" / ..)
            .and(warp::method())
            .and(warp::path::full())
            .and(warp::query::<HashMap<String, String>>())
            .and(warp::header::headers_cloned())
            .and(warp::body::bytes())
            .and(with_proxy(proxy_server.clone()))
            .and_then(handle_proxy_request);

        // Anthropic proxy route
        let anthropic_route = warp::path!("api" / "anthropic" / ..)
            .and(warp::method())
            .and(warp::path::full())
            .and(warp::query::<HashMap<String, String>>())
            .and(warp::header::headers_cloned())
            .and(warp::body::bytes())
            .and(with_proxy(proxy_server.clone()))
            .and_then(handle_proxy_request);

        // Alibaba proxy route
        let alibaba_route = warp::path!("api" / "alibaba" / ..)
            .and(warp::method())
            .and(warp::path::full())
            .and(warp::query::<HashMap<String, String>>())
            .and(warp::header::headers_cloned())
            .and(warp::body::bytes())
            .and(with_proxy(proxy_server.clone()))
            .and_then(handle_proxy_request);

        // MCP proxy route
        let mcp_route = warp::path!("api" / "mcp-proxy" / ..)
            .and(warp::method())
            .and(warp::path::full())
            .and(warp::query::<HashMap<String, String>>())
            .and(warp::header::headers_cloned())
            .and(warp::body::bytes())
            .and(with_proxy(proxy_server.clone()))
            .and_then(handle_proxy_request);

        // Generic proxy route
        let generic_route = warp::path!("api" / "proxy" / ..)
            .and(warp::method())
            .and(warp::path::full())
            .and(warp::query::<HashMap<String, String>>())
            .and(warp::header::headers_cloned())
            .and(warp::body::bytes())
            .and(with_proxy(proxy_server.clone()))
            .and_then(handle_proxy_request);

        // CORS configuration - allow all origins, methods, and headers
        let cors = warp::cors()
            .allow_any_origin()
            .allow_methods(vec!["GET", "POST", "PUT", "DELETE", "PATCH", "OPTIONS", "HEAD"])
            .allow_headers(vec![
                "content-type",
                "authorization",
                "accept",
                "origin",
                "user-agent",
                "cache-control",
                "x-requested-with",
            ])
            .expose_headers(vec![
                "content-type",
                "content-length",
                "cache-control",
                "x-accel-buffering",
            ])
            .max_age(3600);

        let routes = openai_route
            .or(azure_route)
            .or(google_route)
            .or(anthropic_route)
            .or(alibaba_route)
            .or(mcp_route)
            .or(generic_route)
            .with(cors);

        println!("[Tauri Proxy] Starting proxy server on {}", addr);
        warp::serve(routes).run(addr).await;
        
        Ok(())
    }
}

fn with_proxy(
    proxy: Arc<Mutex<ProxyServer>>,
) -> impl Filter<Extract = (Arc<Mutex<ProxyServer>>,), Error = Infallible> + Clone {
    warp::any().map(move || proxy.clone())
}

async fn handle_proxy_request(
    method: warp::http::Method,
    path: warp::path::FullPath,
    query_params: HashMap<String, String>,
    headers: warp::http::HeaderMap,
    body: Bytes,
    proxy: Arc<Mutex<ProxyServer>>,
) -> Result<warp::http::Response<warp::hyper::Body>, Infallible> {
    // Handle OPTIONS preflight requests FIRST (before any other processing)
    if method == warp::http::Method::OPTIONS {
        println!("[Tauri Proxy] Handling OPTIONS preflight request");
        return Ok(warp::http::Response::builder()
            .status(warp::http::StatusCode::NO_CONTENT)
            .header("Access-Control-Allow-Origin", "*")
            .header("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, PATCH, OPTIONS, HEAD")
            .header("Access-Control-Allow-Headers", "content-type, authorization, accept, origin, user-agent, cache-control, x-requested-with")
            .header("Access-Control-Max-Age", "3600")
            .body(warp::hyper::Body::empty())
            .unwrap());
    }
    
    let proxy = proxy.lock().await;
    
    // Get endpoint from query params
    let endpoint = match query_params.get("endpoint") {
        Some(ep) => ep.clone(),
        None => {
            let json_body = serde_json::to_string(&serde_json::json!({
                "error": "Missing endpoint parameter"
            })).unwrap_or_else(|_| "{}".to_string());
            
            return Ok(warp::http::Response::builder()
                .status(warp::http::StatusCode::BAD_REQUEST)
                .header("content-type", "application/json")
                .header("Access-Control-Allow-Origin", "*")
                .body(warp::hyper::Body::from(json_body))
                .unwrap());
        }
    };

    // Build target URL
    // The endpoint parameter already contains the full URL with query params
    // We should not add additional query params from the proxy request
    let target_url = endpoint.clone();

    println!("[Tauri Proxy] {} {}", method, target_url);
    println!("[Tauri Proxy] Request headers: {:?}", headers.keys().collect::<Vec<_>>());

    // Build request headers
    let mut req_headers = reqwest::header::HeaderMap::new();
    let skip_headers = ["connection", "host", "origin", "referer", "cookie", "accept-encoding"];
    
    for (key, value) in headers.iter() {
        let key_str = key.as_str().to_lowercase();
        if !skip_headers.contains(&key_str.as_str()) 
            && !key_str.starts_with("x-") 
            && !key_str.starts_with("sec-") {
            if let Ok(header_name) = reqwest::header::HeaderName::from_bytes(key.as_str().as_bytes()) {
                if let Ok(header_value) = reqwest::header::HeaderValue::from_bytes(value.as_bytes()) {
                    req_headers.insert(header_name, header_value);
                }
            }
        }
    }

    // Make request
    let req_method = match method.as_str() {
        "GET" => reqwest::Method::GET,
        "POST" => reqwest::Method::POST,
        "PUT" => reqwest::Method::PUT,
        "DELETE" => reqwest::Method::DELETE,
        "PATCH" => reqwest::Method::PATCH,
        "HEAD" => reqwest::Method::HEAD,
        _ => reqwest::Method::GET,
    };

    let mut request = proxy.client
        .request(req_method, &target_url)
        .headers(req_headers);

    if !body.is_empty() {
        request = request.body(body);
    }

    match request.send().await {
        Ok(response) => {
            let status = response.status();
            let mut resp_headers = warp::http::HeaderMap::new();
            
            // Copy response headers (preserve important headers like content-type)
            for (key, value) in response.headers().iter() {
                let key_str = key.as_str().to_lowercase();
                // Skip only problematic headers, but keep content-type, content-length, etc.
                if key_str != "www-authenticate" && key_str != "content-encoding" {
                    if let Ok(header_value) = warp::http::HeaderValue::from_bytes(value.as_bytes()) {
                        resp_headers.insert(key.clone(), header_value);
                    }
                }
            }
            
            // Add CORS headers (comprehensive)
            resp_headers.insert(
                warp::http::header::ACCESS_CONTROL_ALLOW_ORIGIN,
                warp::http::HeaderValue::from_static("*"),
            );
            resp_headers.insert(
                warp::http::header::ACCESS_CONTROL_ALLOW_METHODS,
                warp::http::HeaderValue::from_static("GET, POST, PUT, DELETE, PATCH, OPTIONS, HEAD"),
            );
            resp_headers.insert(
                warp::http::header::ACCESS_CONTROL_ALLOW_HEADERS,
                warp::http::HeaderValue::from_static("content-type, authorization, accept, origin, user-agent, cache-control, x-requested-with"),
            );
            resp_headers.insert(
                warp::http::header::ACCESS_CONTROL_EXPOSE_HEADERS,
                warp::http::HeaderValue::from_static("content-type, content-length, cache-control, x-accel-buffering"),
            );
            resp_headers.insert(
                warp::http::header::ACCESS_CONTROL_MAX_AGE,
                warp::http::HeaderValue::from_static("3600"),
            );
            resp_headers.insert(
                "X-Accel-Buffering",
                warp::http::HeaderValue::from_static("no"),
            );

            // Get response body
            match response.bytes().await {
                Ok(body) => {
                    let mut resp = warp::http::Response::builder()
                        .status(status);
                    
                    // Add all headers
                    for (key, value) in resp_headers.iter() {
                        resp = resp.header(key, value);
                    }
                    
                    Ok(resp.body(warp::hyper::Body::from(body)).unwrap())
                }
                Err(e) => {
                    println!("[Tauri Proxy] Failed to read response body: {}", e);
                    let json_body = serde_json::to_string(&serde_json::json!({
                        "error": "Failed to read response body",
                        "message": e.to_string()
                    })).unwrap_or_else(|_| "{}".to_string());
                    
                    Ok(warp::http::Response::builder()
                        .status(warp::http::StatusCode::BAD_GATEWAY)
                        .header("content-type", "application/json")
                        .body(warp::hyper::Body::from(json_body))
                        .unwrap())
                }
            }
        }
        Err(e) => {
            println!("[Tauri Proxy] Request failed: {}", e);
            println!("[Tauri Proxy] Error details: {:?}", e);
            println!("[Tauri Proxy] Target URL was: {}", target_url);
            
            let error_message = if e.is_timeout() {
                "Request timeout"
            } else if e.is_connect() {
                "Connection failed"
            } else if e.is_request() {
                "Invalid request"
            } else {
                "Unknown error"
            };
            
            let json_body = serde_json::to_string(&serde_json::json!({
                "error": "Proxy request failed",
                "message": e.to_string(),
                "errorType": error_message,
                "targetUrl": target_url
            })).unwrap_or_else(|_| "{}".to_string());
            
            Ok(warp::http::Response::builder()
                .status(warp::http::StatusCode::BAD_GATEWAY)
                .header("content-type", "application/json")
                .header("Access-Control-Allow-Origin", "*")
                .body(warp::hyper::Body::from(json_body))
                .unwrap())
        }
    }
}

File: app/api/proxy.ts
import { NextRequest, NextResponse } from "next/server";

export async function handle(
  req: NextRequest,
  { params }: { params: { path: string[] } },
) {
  console.log("[Proxy Route] 🚀 代理API被调用，params:", params);
  console.log("[Proxy Route] 🚀 请求方法:", req.method);
  console.log("[Proxy Route] 🚀 请求路径:", req.nextUrl.pathname);

  if (req.method === "OPTIONS") {
    return NextResponse.json({ body: "OK" }, { status: 200 });
  }

  // 从URL参数中获取endpoint
  const endpoint = req.nextUrl.searchParams.get("endpoint");
  if (!endpoint) {
    return NextResponse.json(
      { error: "Missing endpoint parameter" },
      { status: 400 },
    );
  }

  // remove path params from searchParams
  req.nextUrl.searchParams.delete("path");
  req.nextUrl.searchParams.delete("provider");
  req.nextUrl.searchParams.delete("endpoint");

  // 直接使用endpoint参数作为目标URL
  // endpoint已经包含了完整的API URL（包括baseUrl和path）
  const remainingParams = req.nextUrl.searchParams.toString();
  const fetchUrl = remainingParams
    ? `${endpoint}${endpoint.includes("?") ? "&" : "?"}${remainingParams}`
    : endpoint;

  console.log("[Proxy] Fetching URL:", fetchUrl);
  console.log("[Proxy] Original endpoint:", endpoint);
  console.log("[Proxy] Method:", req.method);

  const skipHeaders = [
    "connection",
    "host",
    "origin",
    "referer",
    "cookie",
    "accept-encoding", // 不转发accept-encoding，让fetch自动处理
  ];
  const headers = new Headers(
    Array.from(req.headers.entries()).filter((item) => {
      if (
        item[0].indexOf("x-") > -1 ||
        item[0].indexOf("sec-") > -1 ||
        skipHeaders.includes(item[0])
      ) {
        return false;
      }
      return true;
    }),
  );
  // 纯前端应用，用户需要提供自己的API密钥

  const controller = new AbortController();
  const fetchOptions: RequestInit = {
    headers,
    method: req.method,
    body: req.body,
    // to fix #2485: https://stackoverflow.com/questions/55920957/cloudflare-worker-typeerror-one-time-use-body
    redirect: "manual",
    // @ts-ignore
    duplex: "half",
    signal: controller.signal,
  };

  const timeoutId = setTimeout(
    () => {
      controller.abort();
    },
    10 * 60 * 1000,
  );

  try {
    const res = await fetch(fetchUrl, fetchOptions);

    // 复制响应头
    const newHeaders = new Headers(res.headers);
    newHeaders.delete("www-authenticate");
    newHeaders.set("X-Accel-Buffering", "no");

    // 检查响应的content-encoding
    const contentEncoding = res.headers.get("content-encoding");

    // 如果响应被压缩，fetch API会自动解压
    // 但我们需要删除content-encoding header，因为返回的body已经是解压后的
    if (contentEncoding) {
      newHeaders.delete("content-encoding");
    }

    // 对于流式响应，需要特殊处理以避免连接中断错误
    if (res.body) {
      const reader = res.body.getReader();
      const stream = new ReadableStream({
        async start(controller) {
          try {
            while (true) {
              const { done, value } = await reader.read();
              if (done) {
                controller.close();
                break;
              }
              controller.enqueue(value);
            }
          } catch (error) {
            // 流错误通常是因为连接被对方关闭，这在流式响应结束时是正常的
            // 只记录非预期的错误
            const errorMessage =
              error instanceof Error ? error.message : String(error);
            if (
              !errorMessage.includes("terminated") &&
              !errorMessage.includes("closed")
            ) {
              console.error("[Proxy] Unexpected stream error:", error);
            }
            // 静默关闭控制器
            try {
              controller.close();
            } catch (e) {
              // 控制器可能已经关闭，忽略错误
            }
          }
        },
        cancel() {
          try {
            reader.cancel();
          } catch (e) {
            // 忽略取消错误
          }
        },
      });

      return new Response(stream, {
        status: res.status,
        statusText: res.statusText,
        headers: newHeaders,
      });
    }

    // 如果没有body，直接返回
    return new Response(null, {
      status: res.status,
      statusText: res.statusText,
      headers: newHeaders,
    });
  } catch (error) {
    console.error("[Proxy] Fetch error:", error);
    console.error("[Proxy] Target URL:", fetchUrl);

    // 返回错误响应而不是抛出异常
    return NextResponse.json(
      {
        error: "Proxy request failed",
        message: error instanceof Error ? error.message : String(error),
        targetUrl: fetchUrl,
      },
      { status: 502 },
    );
  } finally {
    clearTimeout(timeoutId);
  }
}
